pyomo/contrib/solver/ipopt.py

```diff
@@ -0,0 +1,569 @@
+from pyomo.contrib.solver.solution import SolSolutionLoader
+from pyomo.common.tee import TeeStream
+from pyomo.core.expr.visitor import replace_expressions
+from pyomo.core.expr.numvalue import value
+from pyomo.core.base.suffix import Suffix
+from pyomo.common.collections import ComponentMap
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class IpoptSolverError(PyomoException):
+    """
+    General exception to catch solver system errors
+    """
+
+
+class IpoptConfig(SolverConfig):
+    def __init__(
+        self,
+        description=None,
+        doc=None,
+        implicit=False,
+        implicit_domain=None,
+        visibility=0,
+    ):
+        super().__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+
+        self.executable: Executable = self.declare(
+            'executable',
+            ConfigValue(
+                default=Executable('ipopt'),
+                description="Preferred executable for ipopt. Defaults to searching the "
+                "``PATH`` for the first available ``ipopt``.",
+            ),
+        )
+        self.writer_config: ConfigDict = self.declare(
+            'writer_config', NLWriter.CONFIG()
+        )
+
+
+class IpoptResults(Results):
+    def __init__(
+        self,
+        description=None,
+        doc=None,
+        implicit=False,
+        implicit_domain=None,
+        visibility=0,
+    ):
+        super().__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+        self.timing_info.ipopt_excluding_nlp_functions: Optional[float] = (
+            self.timing_info.declare(
+                'ipopt_excluding_nlp_functions',
+                ConfigValue(
+                    domain=NonNegativeFloat,
+                    default=None,
+                    description="Total CPU seconds in IPOPT without function evaluations.",
+                ),
+            )
+        )
+        self.timing_info.nlp_function_evaluations: Optional[float] = (
+            self.timing_info.declare(
+                'nlp_function_evaluations',
+                ConfigValue(
+                    domain=NonNegativeFloat,
+                    default=None,
+                    description="Total CPU seconds in NLP function evaluations.",
+                ),
+            )
+        )
+        self.timing_info.total_seconds: Optional[float] = self.timing_info.declare(
+            'total_seconds',
+            ConfigValue(
+                domain=NonNegativeFloat,
+                default=None,
+                description="Total seconds in IPOPT. NOTE: Newer versions of IPOPT (3.14+) "
+                "no longer separate timing information.",
+            ),
+        )
```

Comment:
Do these need to be declared at all?  I thought the point of `timing_info` was that it allowed implicit definitions and solvers could return whatever hey wanted.